### PR TITLE
sqlstats: fix insights memleak when disabling `sql.metrics.transaction_details.enabled`

### LINKED ISSUE #117300: pkg/sql/sqlstats: disabling `sql.metrics.transaction_details.enabled` prevents locking registry from draining session buffer
**Describe the problem**

Within the SQL Stats subsystem, both statements and transactions are eventually recorded to a [session-specific buffer](https://github.com/cockroachdb/cockroach/blob/9866725b89aeae678f82833b1fcf6272b3483543/pkg/sql/sqlstats/insights/registry.go#L29) within `lockingRegistry`. 

`lockingRegistry.ObserveStatement` [adds statements to the session buffer](https://github.com/cockroachdb/cockroach/blob/9866725b89aeae678f82833b1fcf6272b3483543/pkg/sql/sqlstats/insights/registry.go#L46), and `lockingRegistry.ObserveTransaction` [drains all of the statements from the session buffer](https://github.com/cockroachdb/cockroach/blob/9866725b89aeae678f82833b1fcf6272b3483543/pkg/sql/sqlstats/insights/registry.go#L106). 

The problem is - whenever we disable the `sql.metrics.transaction_details.enabled` cluster setting, `lockingRegistry.ObserveTransaction` is never called, because we [short-circuit the would-be caller of that function when disabled](https://github.com/cockroachdb/cockroach/blob/78ce9b1b89ce5a8ded8ca72f4d7ddea86d5ad6a3/pkg/sql/sqlstats/ssmemstorage/ss_mem_writer.go#L282-L284). Note however that disabling the `sql.metrics.transaction_details.enabled` cluster setting does **not** prevent the caller from eventually calling `lockingRegistry.ObserveStatement`, meaning that we have elements being produced into the buffer, but nothing is ever draining the buffer.

This creates a memory leak in the SQL Stats system when `sql.metrics.transaction_details.enabled` is disabled.

If `sql.metrics.transaction_details.enabled` is disabled, is there any use in making calls to `ObserveStatement`? Without the buffers being flushed to the `lockingRegistry.sink`, it seems like the recorded statements don't provide any value. So perhaps this setting should disable the insights system altogether.

Jira issue: CRDB-35110

### PR DESCRIPTION
**Pull request by abarganier**
#### Reviewer note: best reviewed commit-wise

Fixes: https://github.com/cockroachdb/cockroach/issues/117300

When we record statements to the insights system, they are cached in a
map within the `lockingRegistry`, keyed by session ID. The gathered
statements remain in the map until a transaction for the same session is
recorded to the insights system, which triggers a flush of the
statements to the underlying insights store.

However, if the `sql.metrics.transaction_details.enabled` cluster
setting is disabled while the `sql.metrics.statement_details.enabled` is
left enabled, a memory leak occurs. This is because disabling the
setting prevents transactions being sent to the insights system while
statements continue to be. Therefore, the cache flush trigger mechanism
is removed from the equation, meaning that statements begin to pile up
in the cache indefinitely.

This patch addresses this issue by preventing statements from being
written to the insights system if `sql.metrics.transaction_details.enabled`
is disabled. When the setting is disabled, we also clear the underlying caches
of any existing data.

Release note (bug fix): A memory leak within the insights system was
found to occur when `sql.metrics.transaction_details.enabled` was
disabled, while leaving `sql.metrics.statement_details.enabled` enabled.
This patch fixes the memory leak by preventing the collection of
further statement & transaction insights when
`sql.metrics.transaction_details.enabled` is disabled.

Epic: CRDB-20791

**Workload benchmarks:**

**YCSB**
|   | Type | ops(total) | ops/sec(cum) | avg(ms) | p50(ms) | p95(ms) | p99(ms) | pMax(ms) |
|--|--|--|--|--|--|--|--|--|
| Before | Result | 1488316 | 24372.7 | 0.8 | 0.3 | 5.5 | 9.4 | 62.9 |
| After | Result | 1478367 | 24452.9 | 0.8 | 0.3 | 5.5 | 9.4 | 58.7 |

**Insights**
|   | Type | ops(total) | ops/sec(cum) | avg(ms) | p50(ms) | p95(ms) | p99(ms) | pMax(ms) |
|--|--|--|--|--|--|--|--|--|
| Before | Result | 658 | 10.6 | 1834.5 | 1409.3 | 5100.3 | 6174.0 | 7247.8 |
| After | Result | 644 | 10.7 | 1850.7 | 1476.4 | 4831.8 | 6173.9 | 6979.3 |

### CODE REVIEW

File: pkg/sql/sqlstats/insights/ingester.go [hunk 1]
```diff
@@ -31,12 +31,26 @@ type concurrentBufferIngester struct {
...
 	eventBufferCh chan *eventBuffer
 	registry      *lockingRegistry
 	running       uint64
 }
 
-var _ Writer = &concurrentBufferIngester{}
+// Clear flushes the underlying buffer, and signals the underlying registry
+// to clear any remaining cached data afterwards. This is a synchronous operation.
+func (i *concurrentBufferIngester) Clear(ctx context.Context) {
+	i.ingest(ctx, i.guard.eventBuffer)
```
Comment (abarganier) [inline pkg/sql/sqlstats/insights/ingester.go:48]: Eh, this is wrong the more I think about it. This should be done within the context of the async task that listens on the eventBufferCh. I'll put up a patch.

Comment (abarganier) [reply pkg/sql/sqlstats/insights/ingester.go]: Done.

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117709)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @koorosh, @maryliag, and @xinhaoz)

___
*[`pkg/sql/sqlstats/insights/insights.go` line 165 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/117709#-NoNyDW72l5fk6H5-0uJ:-NoNyDW72l5fk6H5-0uK:blz1su6) ([raw file](https://github.com/cockroachdb/cockroach/blob/298e9b1782313acf8782828b202323bf29f9775b/pkg/sql/sqlstats/insights/insights.go#L165)):*
> ```Go
> 	// Clear clears the cache of its contents, with no guarantees around flush behavior.
> 	// Data may simply be erased depending on the implementation.
> 	Clear(ctx context.Context)
> ```

Did you consider placing the `Clear` method directly on `Writer`? I don't see the value of this separate interface, though I could be missing something justifying it.
___
*[`pkg/sql/sqlstats/insights/registry.go` line 31 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/117709#-NoNxJtX8DFclLpISZtn:-NoNxJtX8DFclLpISZto:b-chi8ub) ([raw file](https://github.com/cockroachdb/cockroach/blob/298e9b1782313acf8782828b202323bf29f9775b/pkg/sql/sqlstats/insights/registry.go#L31)):*
> ```Go
> type lockingRegistry struct {
> 	mu struct {
> 		syncutil.Mutex
> ```

What necessitated the addition of a mutex here? When I was looking at this code it seems like it is only called from a single goroutine (`concurrentBufferIngester.ingest()`). You seem to have taken pains to make sure `Clear()` is called from that same single goroutine.
___
*[`pkg/sql/sqlstats/insights/ingester_test.go` line 152 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/117709#-NoNvLCYBcFFWYE4aiom:-NoNvLCYBcFFWYE4aion:b-etsf6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/298e9b1782313acf8782828b202323bf29f9775b/pkg/sql/sqlstats/insights/ingester_test.go#L152)):*
> ```Go
> 		// with some data. This sets us up to call Clear() with guaranteed data
> 		// in both, so we can assert afterward that both have been cleared.
> 		ingester.testKnobs.noTimedFlush = true
> ```

This feels a bit hacky to me. Perhaps `ingester.Start()` could take an option to not start the timed flush goroutine. Or an option to configure the timed flush period so that it can be set very high. In additional to a feeling of hackiness, this code is currently a data race since you're potentially reading `noTimedFlush` on a different goroutine while it is being set here. So if you keep this you need to set this variable before calling `ingester.Start()`.

**Review by abarganier — COMMENTED** [review]
TFTR! I agree with your comments. I've been on-call this week, but will get back to this PR once my shift has ended. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117709#-:-NoXgjFe2zCar5WluenJ:btnbskv)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @koorosh, @maryliag, and @xinhaoz)

**Review by abarganier — COMMENTED** [review]
Thanks for the feedback, apologies for the delay with the turnaround. I addressed your comments, RFAL!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117709#-:-NpQwLWC5m69yFZj2CmJ:b-ah5u4r)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @koorosh, @maryliag, @petermattis, and @xinhaoz)

___
*[`pkg/sql/sqlstats/insights/insights.go` line 165 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/117709#-NoNyDW72l5fk6H5-0uJ:-NpQjrW_0qeTpz71vkro:b-olmdm9) ([raw file](https://github.com/cockroachdb/cockroach/blob/298e9b1782313acf8782828b202323bf29f9775b/pkg/sql/sqlstats/insights/insights.go#L165)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Did you consider placing the `Clear` method directly on `Writer`? I don't see the value of this separate interface, though I could be missing something justifying it.
</blockquote></details>

My impression was that a `Writer` doesn't necessarily indicate an underlying cache exists to be cleared, hence the separate interface. However, in reality, the only implementations at play here have underlying caches backing them, so I guess it's not really necessary. I went ahead and just made it part of the `Writer` interface.

If some use case in the future arises where it no longer makes sense, folks can split the interface then if necessary.
___
*[`pkg/sql/sqlstats/insights/registry.go` line 31 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/117709#-NoNxJtX8DFclLpISZtn:-NpQqUct2DldDwvr3z8W:bxo4alm) ([raw file](https://github.com/cockroachdb/cockroach/blob/298e9b1782313acf8782828b202323bf29f9775b/pkg/sql/sqlstats/insights/registry.go#L31)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

What necessitated the addition of a mutex here? When I was looking at this code it seems like it is only called from a single goroutine (`concurrentBufferIngester.ingest()`). You seem to have taken pains to make sure `Clear()` is called from that same single goroutine.
</blockquote></details>

Ah, this was a result of a data race being detected within `TestIngester_Clear`. I was making the test reach into the underlying registry that the `concurrentBufferIngester` was wrapping to ensure that it was cleared from the test goroutine, which was triggering the race detector. In hindsight, we shouldn't be adding a mutex to code solely for the purposes of a test.

Perhaps the correct thing to do is to avoid making those assertions within `TestIngester_Clear`, and instead rely on `TestInsightsRegistry_Clear` to test this.

I went ahead and updated the test accordingly, and removed the mutex. 
___
*[`pkg/sql/sqlstats/insights/ingester_test.go` line 152 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/117709#-NoNvLCYBcFFWYE4aiom:-NpQurx1-I3heZoAw1N-:b-726zim) ([raw file](https://github.com/cockroachdb/cockroach/blob/298e9b1782313acf8782828b202323bf29f9775b/pkg/sql/sqlstats/insights/ingester_test.go#L152)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This feels a bit hacky to me. Perhaps `ingester.Start()` could take an option to not start the timed flush goroutine. Or an option to configure the timed flush period so that it can be set very high. In additional to a feeling of hackiness, this code is currently a data race since you're potentially reading `noTimedFlush` on a different goroutine while it is being set here. So if you keep this you need to set this variable before calling `ingester.Start()`.
</blockquote></details>

Thanks for the callout, I think the start options is the right idea. I went ahead and added a `WithoutTimedFlush` option which we can use in tests when calling `.Start()`. Lmk what you think.

**Review by petermattis — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117709#-:-NpV-YsCCLdUs7oqG0oG:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @abarganier, @dhartunian, @koorosh, @maryliag, and @xinhaoz)

___
*[`pkg/sql/sqlstats/insights/insights.go` line 165 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/117709#-NoNyDW72l5fk6H5-0uJ:-NpUzdGG69ottdAIEdlv:b-2mnz5x) ([raw file](https://github.com/cockroachdb/cockroach/blob/298e9b1782313acf8782828b202323bf29f9775b/pkg/sql/sqlstats/insights/insights.go#L165)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

My impression was that a `Writer` doesn't necessarily indicate an underlying cache exists to be cleared, hence the separate interface. However, in reality, the only implementations at play here have underlying caches backing them, so I guess it's not really necessary. I went ahead and just made it part of the `Writer` interface.

If some use case in the future arises where it no longer makes sense, folks can split the interface then if necessary.
</blockquote></details>

Yeah, we can always refactor it in the future. I've noticed that many of the interfaces in the `sqlstats` area have only a single interface. Java and C++ encourage that behavior because retrofitting an interface can be awkward. But in Go retrofitting an interface is usually very straightforward and so my general advice is to only introduce an interface when there is more than one implementation. (Note, don't do anything in this PR about this).
___
*[`pkg/sql/sqlstats/insights/registry.go` line 218 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/117709#-NpV-AZBDS1C6KN7TnO0:-NpV-AZC0rhgsAHZregm:b-7k9z4a) ([raw file](https://github.com/cockroachdb/cockroach/blob/493f44f95d2011c272b0016c8a93f3a32c24f8fa/pkg/sql/sqlstats/insights/registry.go#L218)):*
> ```Go
> 		sink:     sink,
> 	}
> 	lr.statements = make(map[clusterunique.ID]*statementBuf)
> ```

You can move the creation of the `statements` map back into the `lockingRegistry` struct initialization.

**Review by abarganier — COMMENTED** [review]
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117709#-:-NpV7YOfCmUjx8cfiULQ:b-814g8k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian, @koorosh, @maryliag, @petermattis, and @xinhaoz)

___
*[`pkg/sql/sqlstats/insights/insights.go` line 165 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/117709#-NoNyDW72l5fk6H5-0uJ:-NpV7ebE0pE_q2EALDSX:b-6levxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/298e9b1782313acf8782828b202323bf29f9775b/pkg/sql/sqlstats/insights/insights.go#L165)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Yeah, we can always refactor it in the future. I've noticed that many of the interfaces in the `sqlstats` area have only a single interface. Java and C++ encourage that behavior because retrofitting an interface can be awkward. But in Go retrofitting an interface is usually very straightforward and so my general advice is to only introduce an interface when there is more than one implementation. (Note, don't do anything in this PR about this).
</blockquote></details>

Appreciate the insight, thank you!
___
*[`pkg/sql/sqlstats/insights/registry.go` line 218 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/117709#-NpV-AZBDS1C6KN7TnO0:-NpV7grs79T-uZJAPy03:beclky1) ([raw file](https://github.com/cockroachdb/cockroach/blob/493f44f95d2011c272b0016c8a93f3a32c24f8fa/pkg/sql/sqlstats/insights/registry.go#L218)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You can move the creation of the `statements` map back into the `lockingRegistry` struct initialization.
</blockquote></details>

Thanks for the catch. Done.

**Review by petermattis — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/117709)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian, @koorosh, @maryliag, and @xinhaoz)

**Comment by abarganier** [comment]
bors r=petermattis

**Pull request merged**